add discovery function

neophob opened the pull request:
Hi - here's a first try to expose a discovery function to this lib.

Question, how to do transpile from ts to js? this step might be missing from the package.json. As I cannot compile it, I was unable to test

Resolves: #40

AlCalzone commented:
> Question, how to do transpile from ts to js? this step might be missing from the package.json. As I cannot compile it, I was unable to test

If you have `typescript` installed globally: `tsc`, otherwise `node_modules\.bin\tsc`

neophob commented:
Thanks - will update the PR then. Should I modify the packags.json with the tsc job?

AlCalzone commented:
No need to update the package.json. I'm hesitant to add a build script because `npm build` gets called on `npm install` and I'm not publishing the sources on npm. 

Did it work for you? I cannot find my gateway using this method.

neophob commented:
tsc fails, i'll try to fix that first

AlCalzone commented:
I changed it to
```ts
import * as bonjourPackage from "bonjour";
const bonjour = bonjourPackage();

const MDNS_NAME = "coap";

function findTradfri() {
	return new Promise((resolve, reject) => {

		function serviceUpListener(service: bonjourPackage.Service) {
			if (!service || !service.txt || !service.name.startsWith("gw-")) {
				return;
			}
			const foundDevice = {
				name: service.name,
				version: service.txt,
				addresses: (service as any).addresses,
			};
			mdnsBrowser.stop();
			resolve(foundDevice);
		}
		const mdnsBrowser = bonjour.findOne({ type: MDNS_NAME, protocol: "udp" }, serviceUpListener);
		mdnsBrowser.start();
	});
}

module.exports = findTradfri;
```

AlCalzone commented:
You can test just that module by running `node` and then entering `var find = require("./build/lib/discovery"); find().then(result => console.log(result));`

AlCalzone commented:
So, something **is** coming from the gateway. Not sure why its not getting picked up.
![unbenannt](https://user-images.githubusercontent.com/17641229/37179722-e3f1911a-2326-11e8-850e-8a4ee90fbf5f.PNG)

neophob commented:
I updated the PR, added the missing bonjour tsc definition and updated the tsc config (not that I would really understand what I just did). 

Then I applied your correct import from the snipped above - now it compiles fine! after that I enter the build directory, started node repl and then:

```
> const { Discovery } = require('./index.js');
undefined
> Discovery().then(result => console.log(result));
Promise { <pending> }
> { name: 'gw-a0cc2b6f6b31',
  version: { version: '1.3.14' },
  addresses: [ '192.168.178.58', 'fe80::a2cc:2bff:fe6f:6b31' ] }
```

ole!

AlCalzone commented:
Weird... it's not working for me. I get a bunch of answers when I remove the filters but none of them is the gateway. But I do see an answer in Wireshark - although that comes from an IPv6 address.

Do you know if there's a way to make bonjour listen on IPv6 too?

neophob commented:
strange - i saw one error, the stop call before the promise resolve is not needed (will fix that). however your screenshot looks similar like my result (including a ipv4 and an ipv6 address). 
do you have multiple network interfaces? i‘m testing on a macbook

AlCalzone commented:
I've only got one network interface on my desktop PC, so that's probably not the issue. I'm suspecting my router as the IKEA app also does not detect my gateway. But as I'm seeing a response, there should be a way to recognize that.

I've opened an [issue](https://github.com/watson/bonjour/issues/44) in the bonjour repo, maybe that helps.

AlCalzone commented:
I've done some minor changes to change the PR to my preferences. Please test if it still works for you.

I'll see if I get my issue sorted out or I'll buy another router :D

neophob commented on src/lib/discovery.ts:
```diff
@@ -0,0 +1,29 @@
+import * as bonjourPackage from "bonjour";
+const bonjour = bonjourPackage();
+
+export interface DiscoveredGateway {
```
nice - i wanted to do this as well

neophob replied:
changes LGTM. about your router issue, do you have a setting calling "enable upnp" or similar you can enable? did you try to update its firmware?

AlCalzone replied:
Nah that's a stock router from the german Telekom. I can't even give devices the same fixed IP addresses when DHCP is active.
I'm thinking of returning it in favor of a "real" router.

neophob replied:
here's an example with the latest version:

```
> const { discoverGateway }= require('./index.js');
undefined
> discoverGateway().then(result => console.log(result));
Promise { <pending> }
> { name: 'gw-a0cc2bf8535d',
  version: '1.3.14',
  addresses: [ '10.0.0.94', 'fe80::a2cc:2bff:fef8:535d' ] }
```

AlCalzone replied:
Great! Will merge soon.

neophob commented:
Strange, ci fails with `No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.`. did you see such an error in the past?

AlCalzone commented:
I think we can solve that by deferring the instantiation of the bonjour client. My tests import `"./index"` to get dummy coverage for those lines and that imports bonjour, which might fail on travis.

Before:
```ts
const bonjour = bonjourPackage();
// ...
	const mdnsBrowser = bonjour.findOne(
```

After:
```ts
let bonjour; // I think this has to have the type bonjour.Browser or something
// ...
	if (bonjour == null) bonjour = bonjourPackage();
	const mdnsBrowser = bonjour.findOne(
```

neophob commented:
Ole, good point @AlCalzone - i just lazy initialise bonjour, ci is green. test coverage decreased, to only thing we could do here is splitting the parsing from discovery and test the parsing separate. but I think we could do this later, what you mean?

AlCalzone commented:
Don't worry about the coverage, that test will be relatively simple (by substituting the bonjour module with a fake/mock). I'm already in the process of adding tests and will include that one.

AlCalzone commented:
Just tell me when you're done :)

neophob commented:
DONE ;)

AlCalzone commented:
Released as v0.9.0 - thanks a lot!
